oceanspy/compute.py

```diff
@@ -0,0 +1,1776 @@
+    """
+    Compute volume weighted mean.
+
+    Parameters
+    ----------
+    od: OceanDataset
+        oceandataset used to compute 
+    varNameList: 1D array_like, str
+        List of variables (strings). 
+    
+    Returns
+    -------
+    ds: xarray.Dataset 
+        Contains averaged variable named as varName_vw_mean
+    
+    Examples
+    --------
+    >>> od = ospy.open_oceandataset.EGshelfIIseas2km_ASR()
+    >>> ospy.compute.volume_weighted_mean(od, 'Temp')
+    <xarray.Dataset>
+    Dimensions:       (time: 1464)
+    Coordinates:
+      * time          (time) datetime64[ns] 2007-09-01 ... 2008-08-31T18:00:00
+    Data variables:
+        Temp_vw_mean  (time) float64 dask.array<shape=(1464,), chunksize=(40,)>
+    """
+    
+    # Check input
+    if not isinstance(od, _ospy.OceanDataset):
+        raise TypeError('`od` must be OceanDataset')
+        
+    varNameList = _np.asarray(varNameList, dtype='str')
+    if varNameList.ndim == 0: varNameList = varNameList.reshape(1)
+    elif varNameList.ndim >1: raise TypeError('Invalid `varNameList`')
+    
+    # Check grid axis
+    axis_error = [axis for axis in ['X', 'Y', 'Z'] if axis not in od._grid.axes]
+    if len(axis_error)!=0: raise ValueError('{} must be in grid_coords'.format(axis_error))
+    
+    # Add missing variables
+    varList = ['HFacC', 'HFacW', 'HFacS', 'rA', 'rAs', 'rAw', 'rAz', 'drF', 'drC'] + list(varNameList)
+    od = _add_missing_variables(od, varList)
+    
+    # Loop through variables
+    mean = {}
+    for varName in varNameList:
+        # Check dimensions
+        for axis in ['X', 'Y', 'Z']:
+            dims = [od._grid.axes[axis].coords[coord].name for coord in od._grid.axes[axis].coords]
+            if not any(dim in od._ds[varName].dims for dim in dims): 
+                raise ValueError('[{}] must have one of these dimensions {}'.format(varName, dims))
+
+        # Extract HFac
+        if   set(['X', 'Y']).issubset(od._ds[varName].dims):   HFac = od._ds['HFacC']
+        elif set(['Xp1', 'Y']).issubset(od._ds[varName].dims): HFac = od._ds['HFacW']
+        elif set(['X', 'Yp1']).issubset(od._ds[varName].dims): HFac = od._ds['HFacS']
+
+        # Extract rA
+        for Aname in ['rA', 'rAs', 'rAw', 'rAz']:
+            if set(od._ds[Aname].dims).issubset(od._ds[varName].dims): rA = od._ds[Aname]
+
+        # Extract dr
+        for rName in ['drF', 'drC']:
+            if set(od._ds[rName].dims).issubset(od._ds[varName].dims):  dr = od._ds[rName]
+        for coord in od._grid.axes['Z'].coords:
+            if od._grid.axes['Z'].coords[coord].name in od._ds[varName].dims and coord!='center':
+                HFac = od._grid.interp(HFac, 'Z', to=coord, boundary='fill', fill_value=_np.nan)
+                if coord!='outer':
+                    dr = od._grid.interp(od._ds['drF'], 'Z', to=coord, boundary='fill', fill_value=_np.nan)
+
+        # Compute Volume
+        cellVol = dr * HFac * rA
```

Comment:
@ThomasHaine suggested to add a function to compute cell volumes. It's already done by volume weighted means, so we just have to create it's own function. The only tricky part is decide the syntax for volumes on different staggered grids.

Reply:
BTW, there is a bug 2 lines below:
```
# Add and clear
mean[varName+'_vw_mean'] = (od._ds[varName]*cellVol).sum(cellVol.dims) / cellVol.sum()
del cellVol, dr, HFac, rA
```
All cells are summed in the denominator. If sea values are masked (e.g., I want the mean of T>0) then it would give wrong results!